From 3f81620f30664171969ef4d394dd046d04d42c1b Mon Sep 17 00:00:00 2001
From: Devi Priya <quic_devipriy@quicinc.com>
Date: Wed, 5 Jul 2023 15:19:46 +0530
Subject: [PATCH 379/500] usb: dwc3-qcom: fix clock handling in
 dwc3_qcom_remove

In dwc3_qcom_remove, dwc3_qcom_suspend is being called from
pm_runtime_allow function, after switching off the clocks.
In dwc3_qcom_suspend, qscratch region is being accessed.
This results in a crash due to access of un-clocked region.
So, moved the clock off logic after all the PM calls.

Also defer from calling clk_disable_unprepare if is_suspended
is set, since clocks are disabled during dwc3_qcom_suspend.

Change-Id: I75f624f951a872e68c593e588e7d8ee92c52f970
Signed-off-by: Balaji Prakash J <bjagadee@codeaurora.org>
(cherry picked from commit 17b5954738cd2a08843a89cee03529c5ce50f452)
Signed-off-by: Devi Priya <quic_devipriy@quicinc.com>
Signed-off-by: Sridharan S N <quic_sridsn@quicinc.com>
---
 drivers/usb/dwc3/dwc3-qcom.c | 13 +++++++------
 1 file changed, 7 insertions(+), 6 deletions(-)

diff --git a/drivers/usb/dwc3/dwc3-qcom.c b/drivers/usb/dwc3/dwc3-qcom.c
index 4830e36a4f51..c3df950f15ec 100644
--- a/drivers/usb/dwc3/dwc3-qcom.c
+++ b/drivers/usb/dwc3/dwc3-qcom.c
@@ -1002,12 +1002,6 @@ static void dwc3_qcom_remove(struct platform_device *pdev)
 	else
 		platform_device_put(pdev);
 
-	for (i = qcom->num_clocks - 1; i >= 0; i--) {
-		clk_disable_unprepare(qcom->clks[i]);
-		clk_put(qcom->clks[i]);
-	}
-	qcom->num_clocks = 0;
-
 	dwc3_qcom_interconnect_exit(qcom);
 	reset_control_assert(qcom->resets);
 
@@ -1022,6 +1016,13 @@ static void dwc3_qcom_remove(struct platform_device *pdev)
 
 	pm_runtime_allow(dev);
 	pm_runtime_disable(dev);
+
+	for (i = qcom->num_clocks - 1; i >= 0; i--) {
+		if (!qcom->is_suspended)
+			clk_disable_unprepare(qcom->clks[i]);
+		clk_put(qcom->clks[i]);
+	}
+	qcom->num_clocks = 0;
 }
 
 static int __maybe_unused dwc3_qcom_pm_suspend(struct device *dev)
-- 
2.34.1

